File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
      
         
        vector<vector<pair<int,int>>>adj(n);
        
        
        for(auto i : flights)
            adj[i[0]].push_back({i[1], i[2]});
        
        
        queue<pair<int, pair<int,int>>>q;
        
        q.push({0,{src, 0}});
        vector<int>dist(n, 1e7);
        dist[src]=0;
        
        while(!q.empty())
        {
            auto it = q.front();
            q.pop();
            int stops = it.first;
            int node = it.second.first;
            int cost = it.second.second;
            if(stops>k)
                continue;
            
            for(auto i : adj[node])
            {
               if(i.second + cost < dist[i.first])
               {
                   dist[i.first] = i.second+cost;
                   q.push({stops+1, {i.first, dist[i.first]}});
               }
            }
        }
        
        return dist[dst]==1e7 ? -1 : dist[dst];        
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
import queue
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
              
        adj = [[] for _ in range(n)] 
        for flight in flights:
            adj[flight[0]].append((flight[1], flight[2]))
                
        dist = [float('inf')]*n
        q = queue.Queue()        
        q.put((0, src, 0))
        
        while not q.empty():        
            stops, city, cost = q.get()            
        
            if stops > k:
                break
            
            for nextCity, nextCost in adj[city]:
                
                if cost + nextCost < dist[nextCity]:
                    dist[nextCity] = cost + nextCost
                    q.put((stops + 1, nextCity, cost + nextCost))
        
        return dist[dst] if dist[dst] != float('inf') else -1
                
        